File: 3.longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        N = len(s)
        if (N == 0):
            return 0
        mem = set()
        result = 0
        L = R = 0
        while (R < N):
            while (R < N) and (s[R] not in mem):
                mem.add(s[R])
                R += 1
            result = max(result, R - L)

            if (R < N):
                while (s[R] in mem):
                    mem.remove(s[L])
                    L += 1

        return result

File: 3.longest-substring-without-repeating-characters/solution.ts
function lengthOfLongestSubstring(s: string): number {
  let l = 0;
  const appearsInWindow = new Set<string>();
  let result = 0;
  for (let r = 0; r < s.length; r++) {
      const char = s[r];
      if (!appearsInWindow.has(char)) {
          appearsInWindow.add(char);
          continue;
      }
      result = Math.max(result, r - l);
      while (appearsInWindow.has(char)) {
          appearsInWindow.delete(s[l]);
          l++;
      }
      appearsInWindow.add(char);
  }
  result = Math.max(result, s.length - l);
  return result;
};